stanford_soe_r25.module

```diff
@@ -0,0 +1,889 @@
+    'cache' => DRUPAL_CACHE_PER_PAGE,
+    'visibility' => BLOCK_VISIBILITY_LISTED,
+    'pages' => 'r25/*/calendar',
+  );
+  return $blocks;
+}
+
+// function that checks if the current user can book a room, based on
+// room machine_id and how the room is authenticated
+function _stanford_soe_r25_can_book($room_id) {
+
+  // default return array; user is an internal (Drupal) user
+  // who can not book the room
+  $can_book = array(
+    'can_book' => FALSE,
+    'auth' => 'internal',
+    'external_module' => '',
+    'external_acct' => FALSE
+  );
+  if (!empty($room_id)) {  // only check if we have a room id, obviously
+    $rooms = _stanford_soe_r25_room_config_load();   // load my room config array
+    if (!empty($rooms[$room_id]['authentication'])) {   // only continue if room has an auth type
+      // if the room uses internal Drupal accounts, simply check if the current user has the permission
+      if (($rooms[$room_id]['authentication'] == stanford_soe_r25_AUTH_DRUPAL_ACCT ||
+          $rooms[$room_id]['authentication'] == stanford_soe_r25_AUTH_EITHER_ACCT) &&
+        user_access('book r25 rooms')
+      ) {
+        $can_book['can_book'] = TRUE;
+      }
+      else {
+        // if the user can't book by Drupal permission, and the room supports
+        // external accounts, then check the user that way
+        if ($rooms[$room_id]['authentication'] == stanford_soe_r25_AUTH_EXTERN_ACCT ||
+          $rooms[$room_id]['authentication'] == stanford_soe_r25_AUTH_EITHER_ACCT
+        ) {
+          $can_book['auth'] = 'external';
+          // see if any module implements hook_stanford_soe_r25_external_user
+          $externs = module_implements('stanford_soe_r25_external_user');
+          if (!empty($externs) && is_array($externs)) {
+            // if so, just use the first one returned
+            $can_book['external_module'] = $externs[0];
+            // call the stanford_soe_r25_external_user hook for the module found
+            // it will return an array of user contact info if okay, or false if not
+            $external_acct = module_invoke($externs[0], 'stanford_soe_r25_external_user');
+            if (!empty($external_acct)) {
+              // we got back a non-empty array, so assume an authenticated user who can book the room
+              $can_book['can_book'] = TRUE;
+              $can_book['external_acct'] = $external_acct;
+            }
+          }
+        }
+      }
+    }
+  }
+  return $can_book; // return the array defined above
+
+}
+
+/**
+ * implements hook_block_view
+ */
+function stanford_soe_r25_block_view($delta = "") {
+
+  // creates the content of the two blocks defined above and returns a renderable array
+  $block = array();
+
+  // get the room machine id from the current url and see if we can identify a status for it
+  $room_id = NULL;
+  $room_status = stanford_soe_r25_ROOM_STATUS_DISABLED;
+  $rooms = _stanford_soe_r25_room_config_load();
+  $url = explode("/", current_path());
```

Comment:
current_path() is not sanitized and returns the raw $_GET['q'] parameter. Use of the values cannot be trusted without sanitization.